---

### №2 **(6 баллов)**

Выполните задания из файла https://cloud.mail.ru/public/3KMf/xQwX7cbBL. 

---

**УСЛОВИЕ**

В этом задании вам нужно подобрать оптимальное значение k для алгоритма kNN. Будем использовать набор данных Wine, где требуется
предсказать сорт винограда, из которого изготовлено вино, используя
результаты химических анализов.

Выполните следующие шаги:

1. Загрузите выборку Wine по адресу https://archive.ics.uci.edu/ml/machinelearning-databases/wine/wine.data

2. Извлеките из данных признаки и классы. Класс записан в первом
столбце (три варианта), признаки — в столбцах со второго по последний. Более подробно о сути признаков можно прочитать по адресу https://archive.ics.uci.edu/ml/datasets/Wine (см. также файл
wine.names, приложенный к заданию)

3. Оценку качества необходимо провести методом кросс-валидации по
5 блокам (5-fold). Создайте генератор разбиений, который перемешивает выборку перед формированием блоков (shuffle=True). Для
воспроизводимости результата, создавайте генератор KFold с фиксированным параметром random_state=42. В качестве меры качества используйте долю верных ответов (accuracy).

4. Найдите точность классификации на кросс-валидации для метода
k ближайших соседей (sklearn.neighbors.KNeighborsClassifier), при
k от 1 до 50. При каком k получилось оптимальное качество? Чему
оно равно (число в интервале от 0 до 1)? Данные результаты и
будут ответами на вопросы 1 и 2.

5. Произведите масштабирование признаков с помощью функции
sklearn.preprocessing.scale. Снова найдите оптимальное k на кроссвалидации.

6. Какое значение k получилось оптимальным после приведения признаков к одному масштабу? Как изменилось значение качества?
Приведите ответы на вопросы 3 и 4.

Если ответом является нецелое число, то целую и дробную часть необходимо разграничивать точкой, например, 0.5. При необходимости округляйте дробную часть до двух знаков.


---

**РЕШЕНИЕ**

In [ ]:
import numpy as np
import pandas as pd

from sklearn.model_selection import KFold, cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import scale

1. Загрузите выборку Wine по адресу https://archive.ics.uci.edu/ml/machinelearning-databases/wine/wine.data

In [3]:
df = pd.read_csv('wine.data.txt', header=None)

2. Извлеките из данных признаки и классы. Класс записан в первом
столбце (три варианта), признаки — в столбцах со второго по последний. Более подробно о сути признаков можно прочитать по адресу https://archive.ics.uci.edu/ml/datasets/Wine (см. также файл
wine.names, приложенный к заданию)

In [4]:
X = df.iloc[:, 1:].values   # признаки (со 2-го по последний столбец)
y = df.iloc[:, 0].values    # класс (1-й столбец)

3. Оценку качества необходимо провести методом кросс-валидации по
5 блокам (5-fold). Создайте генератор разбиений, который перемешивает выборку перед формированием блоков (shuffle=True). Для
воспроизводимости результата, создавайте генератор KFold с фиксированным параметром random_state=42. В качестве меры качества используйте долю верных ответов (accuracy).

In [5]:
kf = KFold(n_splits=5, shuffle=True, random_state=42)

4. Найдите точность классификации на кросс-валидации для метода
k ближайших соседей (sklearn.neighbors.KNeighborsClassifier), при
k от 1 до 50.
    1. При каком k получилось оптимальное качество?

    2. Чему оно равно (число в интервале от 0 до 1)?

Данные результаты и будут ответами на вопросы 1 и 2.

In [6]:
k_values = range(1, 51)
mean_scores = []

for k in k_values:
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, X, y, cv=kf, scoring='accuracy')
    mean_scores.append(scores.mean())

best_k_raw = k_values[int(np.argmax(mean_scores))]
best_score_raw = max(mean_scores)

print("Без масштабирования:")
print("Оптимальное k:", best_k_raw)       # ответ на вопрос 1
print("Лучшая accuracy:", best_score_raw) # ответ на вопрос 2

Без масштабирования:
Оптимальное k: 1
Лучшая accuracy: 0.7304761904761905


5. Произведите масштабирование признаков с помощью функции
sklearn.preprocessing.scale. Снова найдите оптимальное k на кроссвалидации.

In [7]:
# Масштабирование признаков 
X_scaled = scale(X)

# Кросс-валидация kNN для k=1..50, с масштабированием
k_values_scaled = range(1, 51)
mean_scores_scaled = []

for k in k_values_scaled:
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, X_scaled, y, cv=kf, scoring='accuracy')
    mean_scores_scaled.append(scores.mean())

best_k_scaled = k_values_scaled[int(np.argmax(mean_scores_scaled))]
best_score_scaled = max(mean_scores_scaled)

print("\nПосле масштабирования (scale):")
print("Оптимальное k:", best_k_scaled)        # ответ на вопрос 3
print("Лучшая accuracy:", best_score_scaled)  # используется в ответе на вопрос 4



После масштабирования (scale):
Оптимальное k: 29
Лучшая accuracy: 0.9776190476190475


6. Какое значение k получилось оптимальным после приведения признаков к одному масштабу? Как изменилось значение качества?
Приведите ответы на вопросы 3 и 4.

**ОТВЕТ:**
- Оптимальное k: 29
- Лучшая accuracy: 0.9776190476190475